Documentation: Document which file formats exist

BenWiederhake opened the pull request:
I intentionally don't document the format specifics, to prevent this
file from becoming outdated. Hence, only a list with pointers to where
to find details.

linusg commented on Documentation/FileFormats.md:
```diff
@@ -0,0 +1,58 @@
+Requires the metadata-fields `width`, `height`, `scale`, `format` (see `Gfx::BitmapFormat`), and `pitch`, encoded as decimal strings.
+
+The data is encoded according to `Gfx::determine_storage_format(BitmapFormat)`, so either as
+BGRx8888, BGRA8888, RGBA8888, or 8-bit pallette index. Note that the pallette is not transferred.
```
```suggestion
BGRx8888, BGRA8888, RGBA8888, or 8-bit palette index. Note that the palette is not transferred.
```

(https://www.vocabulary.com/dictionary/pallette :^))

linusg commented on Documentation/FileFormats.md:
```diff
@@ -0,0 +1,58 @@
+
+## glyph/x-fonteditor (Clipboard-only)
+
+Requires the metadata-fields, `char` (usually encoded as a string of length 1: the codepoint as a character),
+`width`, and `height` (encoded as a decimal number).
+
+The data is encoded in width times height many bytes, either 0 (clear) or 1 (set).
```
This is a bit vague, and I'd have to look up how it works. Is it possible to give an actual example in textual representation? How are these fields stored?

BenWiederhake replied:
Can you specify which part is vague? Or is it the overall thing? Maybe it helps to phrase it like this?: "The glyph bitmap data is encoded …"

In general, I try to not be too specific about how it works; this document should just be an entry point to learn about which formats exist, and where to learn the details.

linusg replied:
> Can you specify which part is vague?

Whether the metadata is part of the payload or not. I did look at the source, and it's not - just a regular `HashMap` passed to `GUI::Clipboard::the().set_data()`. With that as context, this makes much more sense, so maybe:

> Requires the following fields to be present in the clipboard metadata `HashMap`: ...

?

BenWiederhake replied:
Right, in this document I should not assume that the reader already knows how the clipboard works.

linusg commented on Documentation/FileFormats.md:
```diff
@@ -0,0 +1,58 @@
+
+## text/uri-list (Clipboard-only)
+
+Newline-delimited set of URIs.
```
would also appreciate an example here, e.g.

```suggestion
Newline-delimited set of URIs.

Example:

```
http://serenityos.org/
https://example.com/foo/bar
```
```

(right?)

BenWiederhake replied:
No, but I'll replace it with a better example, and a list of applications that support it.

linusg replied:
oh, URIs, not URLs. Yeah, I remember what this is about now. Proper example would be nice, thanks :)

linusg commented on Documentation/FileFormats.md:
```diff
@@ -0,0 +1,58 @@
+
+## Application File (`*.af` files)
+
+These files are human-readable and are a subset of the INI-format, have no easily detectable filemagic, and define System Menu entries.
```
```suggestion
These files are human-readable and are a subset of the INI-format, have no easily detectable filemagic, and define System Menu entries.

They are stored in [`/res/apps`](../Base/res/apps).
```

BenWiederhake replied:
Eh, alright, I'll add it in, but I would normally assume the user is able to run `git ls-files *.af` by themselves.

linusg replied:
Sure, everyone can do that. But you forget that this file is also viewable online, and hyperlinks are a nice thing :^)
I'd prefer a simple link over navigating there using the GitHub UI.

linusg commented on Documentation/FileFormats.md:
```diff
@@ -0,0 +1,58 @@
+## Font (`*.font` files)
+
+These files contain bitmap definitions of fonts, either varying-width or fixed-width.
+The header definition can be found in `Gfx::FontFileHeader` in `Userland/Libraries/LibGfx/BitmapFont.cpp`.
```
I always like linking files for easy access:

```suggestion
The header definition can be found in `Gfx::FontFileHeader` in [`Userland/Libraries/LibGfx/BitmapFont.cpp`](../Userland/Libraries/LibGfx/BitmapFont.cpp).
```

linusg commented on Documentation/FileFormats.md:
```diff
@@ -0,0 +1,58 @@
+## GUI Markup Language (`*.gml` files)
+
+These files are human-readable, have no easily detectable filemagic, and define GUI interfaces.
+The format is strongly influenced by QML, the Qt Modeling Lange.
```
```suggestion
The format is strongly influenced by QML, the Qt Modeling Language.
```

BenWiederhake replied:
It's typos like these that make me quest my sanity ;)

linusg commented on Documentation/FileFormats.md:
```diff
@@ -0,0 +1,58 @@
+
+The various services each have their own formats, all very similar, and automatically implemented through LibIPC. The specifics depend on the corresponding source `.ipc` file.
+
+The format can be identified by the format magic, which is derived in `Meta/Lagom/Tools/CodeGenerators/IPCCompiler/main.cpp`
```
```suggestion
The format can be identified by the format magic, which is derived in [`Meta/Lagom/Tools/CodeGenerators/IPCCompiler/main.cpp`](../Meta/Lagom/Tools/CodeGenerators/IPCCompiler/main.cpp)
```

linusg reviewed (commented):
Great initiative!

BenWiederhake commented:
Changes since initial push:
- Added nice "see also"s with pointers to the right files/directories
- Added examples and fixed typos as suggested by linusg (thanks!)
- Rebased

BenWiederhake commented:
Changes since last push:
- Documented rough outline of how clipboard and drag & drop work, so that the clipboard "metadata" makes more sense.
- Rebased
